fix Maven bare ‘SNAPSHOT’ version string

cldvr opened the pull request:
Fixes #5613

### Context
Maven compatibility bug fix, since it allows bare ’SNAPSHOT’ as the version string. Discussed in issue #5613.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

floscher commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/MavenResolver.java:
```diff
@@ -285,7 +285,7 @@ private static boolean isUniqueSnapshot(ModuleComponentIdentifier id) {
 
     protected static boolean isNonUniqueSnapshot(ModuleComponentIdentifier moduleComponentIdentifier) {
-        return moduleComponentIdentifier.getVersion().endsWith("-SNAPSHOT");
+        return moduleComponentIdentifier.getVersion().endsWith("SNAPSHOT");
```
```suggestion
        return moduleComponentIdentifier.getVersion().endsWith("-SNAPSHOT") || "SNAPSHOT".equals(moduleComponentIdentifier.getVersion());
```
or
```suggestion
        return moduleComponentIdentifier.getVersion().matches("^(.*-)?SNAPSHOT$");
```

cldvr replied:
Done! Checking SNAPSHOT equality first for better performance.

melix commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/MavenResolver.java:
```diff
@@ -285,7 +285,8 @@ private static boolean isUniqueSnapshot(ModuleComponentIdentifier id) {
 
     protected static boolean isNonUniqueSnapshot(ModuleComponentIdentifier moduleComponentIdentifier) {
-        return moduleComponentIdentifier.getVersion().endsWith("-SNAPSHOT");
+        return "SNAPSHOT".equals(moduleComponentIdentifier.getVersion()) ||
```
:x: The test is redundant. If `equals` work, then `endsWith` will too. We should avoid too many tests because this code is very sensitive to performance.

cldvr replied:
Hi @melix! Sorry, it was a silly mistake... I will submit an update.

My previous exchange with @floscher resulted in the agreement that allowing only literal `SNAPSHOT` or prefix `-SNAPSHOT` should suffice, forgoing some unusual scenarios and resulting in simpler implementation.

Indeed, performance was a major concern when adding the `equals` check, against using a regular expression instead.

melix commented:
Would be nice if you could rebase your PR against latest `master` so that we can have a better idea of what needs to be fixed, it's currently failing, thanks!

cldvr commented:
Fixed my silly mistake and submitted an update.

> Would be nice if you could rebase your PR against latest `master` so that we can have a better idea of what needs to be fixed, it's currently failing, thanks!

Rebased branch `bug/5613-maven-snapshot` onto `master`, while eliminating intermediate commits.

All tests pass when executing `./gradlew :dependencyManagement:check` on current branch after rebase.

melix commented:
Thanks again for the patch. I'm still a bit worried that the double test is going to slow down resolution. Also we should be consistent everywhere. There are more places where the literal `-SNAPSHOT` is used:

- `org.gradle.api.internal.artifacts.ivyservice.ivyresolve.parser.MavenVersionUtils`, used to infer the status from the version number, I think this one is ok
-  `org.gradle.api.internal.artifacts.repositories.DefaultMavenRepositoryContentDescriptor`, which is used to provide a way to resolve snapshots from a specific Maven repository (content filtering). This one needs to be fixed too.
- `org.gradle.api.internal.artifacts.repositories.resolver.MavenResolver#isNonUniqueSnapshot` that you fixed

So it would be great to check that:

- we can resolve `SNAPSHOT` from a snapshot repository
- publishing a library with version `SNAPSHOT` generates the right module metadata (both .pom and .module), in particular that we can resolve **after** publication of such a library (snapshots are particularly painful during resolution)
- `SNAPSHOT` versions are cached similarly to `-SNAPSHOT` versions

This is a bit cumbersome but that's probably the price to pay for supporting this odd pattern :)

cldvr commented:
> I'm still a bit worried that the double test is going to slow down resolution.

Of course a second check has a cost... However, I don't see a way to optimize it further! Even without doing an actual benchmark I'm pretty sure that replacing the double check with a (pre-compiled) regular expression will yield higher cost.

> Also we should be consistent everywhere. There are more places where the literal `-SNAPSHOT` is used:

Probably it would be better to discuss the approach before jumping to implementation. I will have a look and propose something on issue #5613 to gather feedback.

> So it would be great to check that:
> 
> * we can resolve `SNAPSHOT` from a snapshot repository
> * publishing a library with version `SNAPSHOT` generates the right module metadata (both .pom and .module), in particular that we can resolve **after** publication of such a library (snapshots are particularly painful during resolution)
> * `SNAPSHOT` versions are cached similarly to `-SNAPSHOT` versions

I've previously [commented in the issue](https://github.com/gradle/gradle/issues/5613#issuecomment-597295548) that `SNAPSHOT` dependency resolution from a snapshot repository is working, despite a minor discrepancy in relation to default Maven behaviour.

I have not tested the publication of a SNAPSHOT artefact.

Concerning caching, I've noticed that SNAPSHOT dependency is only retrieved from online repository the first time the build is executed. I guess dependency caching is working properly...

Are there some existing integration tests focused on these scenarios that can be extended to verify Gradle's behaviour with the literal `SNAPSHOT` version string?

melix commented:
Sorry if I wasn't clear, @cldvr, but basically what I'm saying is that this PR misses test coverage. You can take a look at the following tests for inspiration:

- `org.gradle.integtests.resolve.maven.MavenSnapshotResolveIntegrationTest`, tests the resolution of unique and non unique snapshots from Maven, should be updated to include a test with `SNAPSHOT`
- `NonUniqueSnapshotVersionsReadOnlyCacheDependencyResolutionTest` and `SnapshotVersionsReadOnlyCacheDependencyResolutionTest` test the behavior of snapshots in the context of the read-only metadata cache
- `org.gradle.api.publish.maven.MavenPublishSnapshotIntegTest` tests the publication of snapshots and verify that we can resolve published snapshot correctly

Our performance test suite will take care of checking if your added test will impact resolution badly or not.
